server/src/test/java/org/elasticsearch/indices/recovery/IndexRecoveryIT.java

```diff
@@ -850,16 +846,24 @@ public void messageReceived(TransportRequestHandler<TransportRequest> handler, T
         }
     }
 
-    private static class SingleStartEnforcer implements Consumer<String> {
+    private class SingleStartEnforcer implements BiConsumer<String, TransportRequest> {
 
         private final AtomicBoolean recoveryStarted = new AtomicBoolean(false);
+        private final String indexName;
+
+        private SingleStartEnforcer(String indexName) {
+            this.indexName = indexName;
+        }
 
         @Override
-        public void accept(String action) {
+        public void accept(String action, TransportRequest request) {
             // The cluster state applier will immediately attempt to retry the recovery on a cluster state
             // update. We want to assert that the first and only recovery attempt succeeds
             if (PeerRecoverySourceService.Actions.START_RECOVERY.equals(action)) {
-                if (recoveryStarted.compareAndSet(false, true) == false) {
+                StartRecoveryRequest startRecoveryRequest = (StartRecoveryRequest) request;
+                ShardId shardId = startRecoveryRequest.shardId();
+                logger.info("--> attempting to send start_recovery request for shard: " + shardId);
+                if (indexName.equals(shardId.getIndexName()) && recoveryStarted.compareAndSet(false, true) == false) {
```

Comment:
I can reproduce this by inserting`Thread.sleep(Randomness.get().nextInt(100))` before we [apply](https://github.com/elastic/elasticsearch/blob/95a7eed9aa35f47b228e402508709b5bd6703cf4/server/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java#L215) the cluster state.

Peer recovery can start more than once if the recovery source hasn't applied the [latest](https://github.com/elastic/elasticsearch/blob/4ed0dc8703ffdb5c3a3f52cbb3ec0f616080aba1/server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java#L164) cluster state. Should we only set the flag `recoveryStarted` after we have sent/received any retryable action?